File: DiscoverTheOntime.py
#! /usr/bin/env python

#
# DiscoverTheOntime.py --- Discover the ontime with users from ASA syslog
#
import os
import sys
import re
import itertools
import datetime
from optparse import OptionParser
import json

usage = "usage: %prog [options] 3> [unterminated log file]"

def fromISO8601ExtDateTime(text):
    sep = text.split ('+')
    if sep[1] != "09:00":
        raise ValueError, "This program only timezone +09:00"
    else:
        result = datetime.datetime.strptime (sep[0], "%Y-%m-%dT%H:%M:%S")
        return result

def extract_start_session_from_log(line, timeField, ipField, classField, messageField):
    result = {}
    result["terminate log"] = None
    result["line"]          = line
    #
    result["Message"]  = messageField
    result["Class"]    = classField
    result["DateTime"] = fromISO8601ExtDateTime (timeField)
    #
    matches = re.match (r'Group <(.*?)> User <(.*?)> IP <(.*?)>', messageField)
    result["Group"]    = matches.group(1)
    result["User"]     = matches.group(2)
    result["IP"]       = matches.group(3)
    return result

def extract_terminate_session_from_log(line, timeField, ipField, classField, messageField):
    result = {}
    result["start log"]     = None
    result["line"]          = line
    #
    result["Message"]  = messageField
    result["Class"]    = classField
    result["DateTime"] = fromISO8601ExtDateTime (timeField)
    #
    pattern = r'Group = (.*?), Username = (.*?), IP = (.*?), Session disconnected\. Session Type: (.*?), Duration: (.*?), '
    matches = re.match (pattern, messageField)
    result["Group"]         = matches.group(1)
    result["User"]          = matches.group(2)
    result["IP"]            = matches.group(3)
    result["SessionType"]   = matches.group(4)
    result["Duration"]      = matches.group(5)
    return result


#
# entry point
#
parser = OptionParser (usage)

parser.add_option(
    "-y", "--year",
    action="store",
    type="int",
    default= datetime.datetime.today ().year,
    dest="year",
    help="set year of log date, as default is this year."
)

(options, args) = parser.parse_args()

#
# main loop
#
start_session_logs = []
terminate_session_logs = []

for line in sys.stdin:
    m = re.match (r'([^ ]+) ?([^ ]+) ?(%ASA-[^:]+?-113039): (.*)$', line)
    if m:
        session = extract_start_session_from_log (line, m.group(1), m.group(2), m.group(3), m.group(4))
        start_session_logs.append (session)
    #
    m = re.match (r'([^ ]+) ?([^ ]+) ?(%ASA-[^:]+?-113019): (.*)$', line)
    if m:
        session = extract_terminate_session_from_log (line, m.group(1), m.group(2), m.group(3), m.group(4))
        terminate_session_logs.append (session)

#
# combine the relationship
#
for start in start_session_logs:
    term = next(itertools.ifilter(lambda it:it["DateTime"] >= start["DateTime"] and it["IP"] == start["IP"], terminate_session_logs), None)
    if term:
        start["terminate log"] = term
        term["start log"] = start

pipR, pipW = os.pipe ()
os.close (pipR)
os.dup2 (3, pipW)
unterminated_file = os.fdopen (pipW, "w")

for item in itertools.ifilter(lambda it:it["terminate log"] == None, start_session_logs):
    print >> unterminated_file, item["line"],
#
for item in itertools.ifilter(lambda it:it["start log"] == None, terminate_session_logs):
    print >> sys.stderr, "This record,"
    print >> sys.stderr, item["line"] ,
    print >> sys.stderr, " has no relationship!"

#
# To output
#
output_logs = []
for it in start_session_logs:
    if it["terminate log"]:
        container = {}
        container.update ({    "start_time": it["DateTime"]})
        container.update ({"terminate_time": it["terminate log"]["DateTime"]})
        container.update ({          "User": it["terminate log"]["User"]})
        container.update ({            "IP": it["terminate log"]["IP"]})
        container.update ({   "SessionType": it["terminate log"]["SessionType"]})
        container.update ({      "Duration": it["terminate log"]["Duration"]})
        output_logs.append (container)
 
def default_datetime(o):
    if isinstance(o, datetime.datetime):
        return o.isoformat()
    raise TypeError(repr(o) + " is not JSON serializable")

json.dump (output_logs, sys.stdout,
        ensure_ascii=True,
        check_circular=True,
        separators=None,
        encoding='utf-8',
        default=default_datetime,
        sort_keys=False,
        indent=4)

File: DiscoverTheOntime.rb
#! /usr/bin/env ruby
#
# DiscoverTheOntime.rb --- Discover the ontime with users from ASA syslog
#
require 'time'

#開始ログ行
re_start_line = /
    (\S+)                          #1 ISOフォーマットの時間
    [^%]+?                         #  現在無視するフィールド
    (%ASA-[^:]+?-113039):          #2 %ASA-6-113039
    \ Group\ <(.*?)>               #3 Group
    \ User\ <(.*?)>                #4 User
    \ IP\ <(.*?)>                  #5 IP
    \ (.*)$                        #6 食い残し
/x

#終了ログ行
re_term_line = /
    (\S+)                          #1 ISOフォーマットの時間
    [^%]+?                         #  現在無視するフィールド
    (%ASA-[^:]+?-113019):          #2 %ASA-4-113019
    \ Group\ =\ (.*?),             #3 Group
    \ Username\ =\ (.*?),          #4 Username
    \ IP\ =\ (.*?),                #5 IP
    \ Session\ disconnected\.
    \ Session\ Type:\ (.*?),       #6 Session Type
    \ Duration:\ (.*?),            #7 Duration
    \ (.*)$                        #8 食い残し
/x

start_session_logs = []
terminate_session_logs = []

#
# log file format error
#
class LogFileFormatError < StandardError; end

#
# main loop
#
STDIN.each_line do |line|
    begin
        if line.include?('-113039:') then
            #開始ログ行を取り出す
            m = re_start_line.match(line) or
                raise LogFileFormatError, "L##{STDIN.lineno} - Sorry, unsupported log format."
            start_session_logs.push({
                :TerminateLog => nil,
                        :Line => line,
                    :DateTime => Time.iso8601(m[1]),
                       :Class => m[2],
                       :Group => m[3],
                        :User => m[4],
                          :IP => m[5],
                    :Leftover => m[6],
            })
        elsif line.include?('-113019:') then
            #終了ログ行を取り出す
            m = re_term_line.match(line) or
                raise LogFileFormatError, "L##{STDIN.lineno} - Sorry, unsupported log format."
            terminate_session_logs.push({
                :StartLog => nil,
                    :Line => line,
                :DateTime => Time.iso8601(m[1]),
                   :Class => m[2],
                   :Group => m[3],
                    :User => m[4],
                      :IP => m[5],
             :SessionType => m[6],
                :Duration => m[7],
                :Leftover => m[8],
            })
        end
    rescue LogFileFormatError => ex
        STDERR.puts ex
    end
end

#
# combine the relationship
#
start_session_logs.each do |start|
    term = terminate_session_logs.find {|it| it[:DateTime] >= start[:DateTime] and it[:IP] === start[:IP]}
    unless term.nil? then
        start[:TerminateLog] = term
        term[:StartLog] = start
    end
end

IO.open( 3, mode="w" ) do |out|
    nopair = start_session_logs.select {|it| it[:TerminateLog].nil?}
    nopair.each do |item|
        out.print item[:Line]
    end
end

nopair = terminate_session_logs.select {|it| it[:StartLog].nil?}
nopair.each do |item|
    STDERR.puts "This record,"
    STDERR.print item[:Line]
    STDERR.puts " has no relationship!"
end

#
# To output
#
buf = []
start_session_logs.each do |item|
    numb = buf.length
    st = item
    tm = item[:TerminateLog]
    unless tm.nil? then
        stime = st[:DateTime].getlocal.strftime('%Y, %m, %d, %H, %M, %S')
        ttime = tm[:DateTime].getlocal.strftime('%Y, %m, %d, %H, %M, %S')
        buf.push("[ '#{numb+1}', '#{st[:User]}', new Date(#{stime}), new Date(#{ttime})],")
    end
end
datarows = buf.join("\n            ")

#
#
#
text = <<"EOS"
<html>
<head>
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<script type="text/javascript" src="https://www.google.com/jsapi"></script>
<script type="text/javascript">
    google.load('visualization', '1', {'packages':['timeline']});
    google.setOnLoadCallback(drawChart);

    function drawChart() {
        var data = new google.visualization.DataTable();
        data.addColumn({ type: 'string', id: 'Number' });
        data.addColumn({ type: 'string', id: 'Name' });
        data.addColumn({ type: 'date', id: 'Start' });
        data.addColumn({ type: 'date', id: 'End' });
        data.addRows([
            #{datarows}
        ]);

        var chart = new google.visualization.Timeline(document.getElementById('chart_div'));

        var options = {
            timeline: {
                showRowLabels: false,
            }
        };

        chart.draw(data, options);
      }
</script>
</head>
<body>
    <div id="chart_div" style="height:100%"></div>
</body>
</html>
EOS
print text unless buf.empty?
